buildSrc/src/main/groovy/org/gradle/testing/JUnitXmlTestEventsGenerator.groovy

```diff
@@ -0,0 +1,140 @@
+        testResult.testcase.each { testCase ->
+            String testCaseClassName = testCase.@classname.text()
+            String testMethodName = testCase.@name.text()
+            def testCaseDescriptor = new DecoratingTestDescriptor(new DefaultTestMethodDescriptor(0, testCaseClassName, testMethodName), testSuiteDescriptor)
+            def skipped = testCase.skipped
+            def failure = testCase.failure
+            long endTime = startTime + (testCase.@time.toDouble() * 1000).round()
+
+            if (failure.size() > 0) {
+                testListener.beforeTest(testCaseDescriptor)
+                publishAdditionalMetadata(testCaseDescriptor, build)
+                try {
+                    String systemErr = testResult."system-err".text()
+                    if (systemErr) {
+                        testOutputListener.onOutput(testCaseDescriptor, new DefaultTestOutputEvent(TestOutputEvent.Destination.StdErr, systemErr))
+                    }
+                } catch (Exception e) {
+                    e.printStackTrace()
+                }
+                String failureText = failure.text()
+                failureText = failureText.replace("java.lang.AssertionError: ", "")
+                testListener.afterTest(testCaseDescriptor, new DefaultTestResult(TestResult.ResultType.FAILURE, startTime, endTime, 1, 0, 1, [new AssertionError(failureText)]))
+            } else if (!(skipped.size() > 0)) {
+                testListener.beforeTest(testCaseDescriptor)
+                publishAdditionalMetadata(testCaseDescriptor, build)
+                testListener.afterTest(testCaseDescriptor, new DefaultTestResult(TestResult.ResultType.SUCCESS, startTime, endTime, 1, 1, 0, []))
+            }
+        }
+        testListener.afterSuite(testSuiteDescriptor, new DefaultTestResult(TestResult.ResultType.SUCCESS, 0, 0, 0, 0, 0, []))
+        testListener.afterSuite(testSuiteDescriptor.parent, new DefaultTestResult(TestResult.ResultType.SUCCESS, 0, 0, 0, 0, 0, []))
+        testListener.afterSuite(testSuiteDescriptor.parent.parent, new DefaultTestResult(TestResult.ResultType.SUCCESS, 0, 0, 0, 0, 0, []))
+    }
+
+    private void publishAdditionalMetadata(DecoratingTestDescriptor testCaseDescriptor, Object build) {
+        def buildUrl = build.@webUrl
+        if (buildUrl) {
+            testOutputListener.onOutput(testCaseDescriptor, new DefaultTestOutputEvent(TestOutputEvent.Destination.StdOut,
+                """
+                    Scenario: ${getScenarioId(build)}
+                    Worker build url: ${buildUrl}
+                """.stripIndent())
+            )
+        }
+    }
+
+    private String getScenarioId(Object build) {
+        NodeChildren properties = build.properties.children()
+        properties.find { it.@name == 'scenario' }.@value.text()
+    }
+
+
+    void addTestListener(TestListener listener) {
+        if (!listener.getClass().getName().startsWith('org.gradle.api.internal')) {
```

Comment:
It is not suppressing the events (that is done by not calling `super.addTestListener`) but ignoring the listeners added by the test task itself - these would be responsible for generating an html and xml report. Even worse, one of the `TestOutputListener`s (`TestReportDataCollector`) would try to write to the `outputWriter` which is closed after the "actual" test execution is finished.